File: python/51-100/52. N-Queens II.py
class Solution(object):

    def checkcol(self, row, col, path):
        for i in range(row):
            if (i, col) in path:
                return False
        return True

    def checkdig(self, row, col, path):
        for j in range(1, row + 1):
            if (row - j, col + j) in path or (row - j, col - j) in path:
                return False
        return True

    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        self.path = {}
        self.res = 0
        def helper(row, n):
            if len(self.path) == n:
                self.res += 1
                return
            for col in range(n):
                if self.checkcol(row, col, self.path) and self.checkdig(row, col, self.path):
                    self.path[(row, col)] = True
                    helper(row + 1, n)
                    del self.path[(row, col)]

        helper(0, n)
        return self.res
sol = Solution()
print sol.totalNQueens(8)

File: cpp/51-100/52. N-Queens II.cpp
class Solution {
public:
    int totalNQueens(int n) {
        int count = 0;
		vector<bool> flags(n, false);
		vector<bool> diag(2 * n - 1, false);
		vector<bool> anti_diag(2 * n - 1, false);
		helper(n, flags, diag, anti_diag, 0, count);
		return count;
    }
    void helper(int n, vector<bool> &flags, vector<bool> &diag, vector<bool> &anti_diag, int level, int &count)
	{
		if (level == n)
		{
			count++;
			return ;
		}
		for (int i = 0; i < n; i++)
		{
			if (flags[i] || diag[n - level - 1 + i] || anti_diag[level + i])
				continue;
			flags[i] = diag[n - level - 1 + i] = anti_diag[level + i] = true;
			helper(n, flags, diag, anti_diag, level + 1, count);
			flags[i] = diag[n - level - 1 + i] = anti_diag[level + i] = false;
		}
	}
};